README.rst
===============================================
 django-celery - Celery Integration for Django
===============================================

.. image:: http://cloud.github.com/downloads/celery/celery/celery_128.png

:Version: 3.2.1
:Web: http://celeryproject.org/
:Download: http://pypi.python.org/pypi/django-celery/
:Source: http://github.com/celery/django-celery/
:Keywords: celery, task queue, job queue, asynchronous, rabbitmq, amqp, redis,
  python, django, webhooks, queue, distributed

--

.. warning::

    **THIS PROJECT IS ONLY REQUIRED IF YOU WANT TO USE DJANGO RESULT BACKEND
    AND ADMIN INTEGRATION**

    Please follow the new tutorial at:

    http://docs.celeryproject.org/en/latest/django/first-steps-with-django.html

django-celery provides Celery integration for Django; Using the Django ORM
and cache backend for storing results, autodiscovery of task modules
for applications listed in ``INSTALLED_APPS``, and more.

Using django-celery
===================

To enable ``django-celery`` for your project you need to add ``djcelery`` to
``INSTALLED_APPS``::

    INSTALLED_APPS += ("djcelery", )

then add the following lines to your ``settings.py``::

    import djcelery
    djcelery.setup_loader()

Everything works the same as described in the `Celery User Manual`_, except you
need to invoke the programs through ``manage.py``:

=====================================  =====================================
**Program**                            **Replace with**
=====================================  =====================================
``celery``                             ``python manage.py celery``
``celery worker``                      ``python manage.py celery worker``
``celery beat``                        ``python manage.py celery beat``
``celery ...``                         ``python manage.py celery ...``
=====================================  =====================================

The other main difference is that configuration values are stored in
your Django projects' ``settings.py`` module rather than in
``celeryconfig.py``.

If you're trying celery for the first time you should start by reading
`Getting started with django-celery`_

Special note for mod_wsgi users
-------------------------------

If you're using ``mod_wsgi`` to deploy your Django application you need to
include the following in your ``.wsgi`` module::

    import djcelery
    djcelery.setup_loader()

Documentation
=============

The `Celery User Manual`_ contains user guides, tutorials and an API
reference. It also has a dedicated `subsection about the Django integration`_.

.. _`Celery User Manual`: http://docs.celeryproject.org/
.. _`subsection about the Django integration`:
   http://docs.celeryproject.org/en/latest/django/
.. _`Getting started with django-celery`:
   http://docs.celeryproject.org/en/latest/django/first-steps-with-django.html

Installation
=============

You can install ``django-celery`` either via the Python Package Index (PyPI)
or from source.

To install using ``pip``,::

    $ pip install django-celery

To install using ``easy_install``,::

    $ easy_install django-celery

You will then want to create the necessary tables. If you generating
schema migrations, you'll want to run::

    $ python manage.py migrate djcelery



Downloading and installing from source
--------------------------------------

Download the latest version of ``django-celery`` from
http://pypi.python.org/pypi/django-celery/

You can install it by doing the following,::

    $ tar xvfz django-celery-0.0.0.tar.gz
    $ cd django-celery-0.0.0
    # python setup.py install # as root

Using the development version
------------------------------

You can clone the git repository by doing the following::

    $ git clone git://github.com/celery/django-celery.git

Getting Help
============

Mailing list
------------

For discussions about the usage, development, and future of celery,
please join the `celery-users`_ mailing list. 

.. _`celery-users`: http://groups.google.com/group/celery-users/

IRC
---

Come chat with us on IRC. The **#celery** channel is located at the `Freenode`_
network.

.. _`Freenode`: http://freenode.net


Bug tracker
===========

If you have any suggestions, bug reports or annoyances please report them
to our issue tracker at http://github.com/celery/django-celery/issues/

Wiki
====

http://wiki.github.com/celery/celery/

Contributing
============

Development of ``django-celery`` happens at Github:
http://github.com/celery/django-celery

You are highly encouraged to participate in the development.
If you don't like Github (for some reason) you're welcome
to send regular patches.

License
=======

This software is licensed under the ``New BSD License``. See the ``LICENSE``
file in the top distribution directory for the full license text.

.. # vim: syntax=rst expandtab tabstop=4 shiftwidth=4 shiftround



Changelog
================
 Change history
================

.. contents::
    :local:

.. _version-3.2.0:

3.2.0
=====
:release-date: TDB

- Now compatible with Django 1.10

- Dropped compatibility with Django 1.7 and earlier.

    For older Django versions use ``django-celery`` 3.1.x

- Prevented redundant migrations when using Python 3.

    Fixes contributed by Ivan Larin and Vytis Banaitis.

- Fixed task cleanup when TaskState has related models.

    Fix contributed by georgewhewell.

- Fixed Django 1.10 compatibility issue in scheduler

    Fix contributed by Mathieu Fenniak.

- Fixed Django 1.10 compatibility issue in management commands

    Fix contributed by Stranger6667, yjmade and Vytis Banaitis.

- Fixed cache backend compatibility issue with Django 1.9

    Fix contributed by Vytis Banaitis, with initial patch by Brandon Craft.

- Fixed missing app label in Django admin.

    Fix contributed by Rockallite Wulf.

- Fix timezone handling in Camera and admin.

    Fix contributed by Vytis Banaitis.

- Added friendly API for dynamically creating/updating/deleting tasks

    Fix contributed by Lev Berman

- Fixed some django 2.0 deprecations.

    Fix contributed by Éloi Rivard

- Resolved some deprecation warnings in the test suite.

    Fix contributed by Vytis Banaitis.

- PeriodicTask admin list view changes:

    - ``task``, ``args``, ``kwargs`` fields added to ``list_display``.

    - Added ability to search tasks by ``name`` and ``task``.

    - Tasks now ordered by ``('-enabled', 'name')``.

    - Added enable/disable actions.

    Contributed by Armenak Baburyan.

.. _version-3.1.17:

3.1.17
======
:release-date: 2015-10-09 04:02 P.M PDT

- Adds Django migrations

    ... warning::

        Old South migrations have been moved to ``djcelery.south_migrations``,
        so if you still use South you need to configure the new location
        in ``settings.py``:

        .. code-block:: python

            SOUTH_MIGRATION_MODULES = {
                'djcelery': 'djcelery.south_migrations',
            }

- ``djcelery.picklefield`` now compatible with Django 1.8

    Fix taken upstream from ``django-picklefield``.

- Fixed timezone issues when CELERY_ENABLE_UTC not set.

    Fix contributed by Dmitriy Krasilnikov.

- Fixed Python 3 compatibility issue in ``setup_loader``
  (Issue #344, Issue #346).

.. _version-3.1.16:

3.1.16
======
:release-date: 2014-09-19 00:00 06:00 P.M UTC

- Fixed a problem with management commands hanging when
  ``djcelery`` is added to ``INSTALLED_APPS``.

.. _version-3.1.15:

3.1.15
======
:release-date: 2014-09-15 00:00 P.M UTC

- Now depends on Celery 3.1.15

- Better Python 3 compatibility

    Contributed by Piotr Bulinski.

- Commands now disables model validation when using Django > 1.7 (Issue #336).

- No longer shadows the new ``--no-color`` option in Django 1.7.

- Commands now have ``.stderr`` and ``.stdout`` attributes for Django 1.7
  compatibility.

    Fix contributed by artscoop.

- ``WorkerState.is_alive()`` is now accurate with ``USE_TZ`` enabled.

    Fix contributed by Rockallite Wulf.

.. _version-3.1.10:

3.1.10
======
:release-date: 2014-03-31 04:00 P.M UTC

- Test suite Django version compatibility fixes

    Contributed by Spencer Ellinor

- Admin: Model forms now defines ``fields`` and ``exclude`` attributes
  (Issue #311)

    Contributed by David Fischer.

- Models: ``get_query_set`` renamed to ``get_queryset`` (but alias available)
  (Issue #308).

    Contributed by David Fischer.

- Admin: Fixed bug in ``humanize.py``.

    Contributed by koodjo.

- Support for Django 1.8's ``DiscoverRunner``.

    Contributed by David Fischer.

- Snapshot: Fixed time stamp timezone issue.

    Contributed by Michael van Tellingen.

.. _version-3.1.9:

3.1.9
=====
:release-date: 2014-02-13 04:35 P.M UTC

- Now depends on Celery 3.1.9

- TestSuiteRunner should now be working again.

- Database result backend now works with Python 3.

- The `celery worker' management command now gives a proper error message
  when positional arguments are provided.

    Fix contributed by Brett Gibson.

- No longer uses manual transaction management to be compatible with Django
  1.6.

    This will probably decrease performance considerably, e.g. for clean up
    of celerycam events and for celerybeat which now needs to commit the
    transaction for every operation.

- The `celery` management command now performs validation checks (Issue #253).

- Url handlers: Fixed compatibility with Django 1.2

    Fix contributed by dongweiming.

.. _version-3.1.1:

3.1.1
=====
:release-date: 2013-11-11 07:30 P.M UTC

- Fixed Python 3 compatibility problems.

    Fix contributed by Anton Novosyolov.

- Fixed compatibility with Django pre 1.6 (Issue #285).

- celerycam: Fixed error when expire function returned None (Issue #210)

- DatabaseScheduler: Now automatically disables periodic task
  if the schedule model it was connected is deleted (Issue #208).

- Test runner now sets always eager settings on the app configuration.
  (Issue #201).

- TaskRunnerStoringResults stored results in the wrong table.

    Fix contributed by Charlie DeTar.

- Celerycam: Fixed problem when USE_TZ not enabled.

    Fix contributed by Jens Alm

- Celerycam: Fixed timezone problem with expires and eta
  (Issue #275 + Issue #153).

    Fix contributed by Alvaro Vega.

- Autodiscovery now warns instead of propagating import
  errors (Issue #226).

.. _version-3.1.0:

3.1.0
=====
:release-date: 2013-11-10 1:04 A.M UTC

.. note::

    Please test this release in staging first and
    make sure you read the Celery 3.1 changelog at:

    http://docs.celeryproject.org/en/latest/whatsnew-3.1.html

    If you upgrade from an old version then you must also
    double check that you have 'djcelery.setup_loader()'
    in your settings.py!

- Tests passing on Django 1.6

- Now works with Celery 3.1

.. _version-3.0.23:

3.0.23
======
:release-date: 2013-09-03 02:00 P.M BST

- Now depends on celery 3.0.23

djcelery/models.py
from __future__ import absolute_import, unicode_literals

from datetime import timedelta
from time import time, mktime, gmtime

from django.core.exceptions import MultipleObjectsReturned, ValidationError
from django.db import models
from django.db.models import signals
from django.utils.translation import ugettext_lazy as _
from django.conf import settings

from celery import schedules
from celery import states
from celery.events.state import heartbeat_expires

from . import managers
from .picklefield import PickledObjectField
from .utils import now
from .compat import python_2_unicode_compatible

ALL_STATES = sorted(states.ALL_STATES)
TASK_STATE_CHOICES = sorted(zip(ALL_STATES, ALL_STATES))


def cronexp(field):
    return field and str(field).replace(' ', '') or '*'


@python_2_unicode_compatible
class TaskMeta(models.Model):
    """Task result/status."""
    task_id = models.CharField(_('task id'), max_length=191, unique=True)
    status = models.CharField(
        _('state'),
        max_length=50, default=states.PENDING, choices=TASK_STATE_CHOICES,
    )
    result = PickledObjectField(null=True, default=None, editable=False)
    date_done = models.DateTimeField(_('done at'), auto_now=True)
    traceback = models.TextField(_('traceback'), blank=True, null=True)
    hidden = models.BooleanField(editable=False, default=False, db_index=True)
    # TODO compression was enabled by mistake, we need to disable it
    # but this is a backwards incompatible change that needs planning.
    meta = PickledObjectField(
        compress=True, null=True, default=None, editable=False,
    )

    objects = managers.TaskManager()

    class Meta:
        verbose_name = _('task state')
        verbose_name_plural = _('task states')
        db_table = 'celery_taskmeta'

    def to_dict(self):
        return {'task_id': self.task_id,
                'status': self.status,
                'result': self.result,
                'date_done': self.date_done,
                'traceback': self.traceback,
                'children': (self.meta or {}).get('children')}

    def __str__(self):
        return '<Task: {0.task_id} state={0.status}>'.format(self)


@python_2_unicode_compatible
class TaskSetMeta(models.Model):
    """TaskSet result"""
    taskset_id = models.CharField(_('group id'), max_length=191, unique=True)
    result = PickledObjectField()
    date_done = models.DateTimeField(_('created at'), auto_now=True)
    hidden = models.BooleanField(editable=False, default=False, db_index=True)

    objects = managers.TaskSetManager()

    class Meta:
        """Model meta-data."""
        verbose_name = _('saved group result')
        verbose_name_plural = _('saved group results')
        db_table = 'celery_tasksetmeta'

    def to_dict(self):
        return {'taskset_id': self.taskset_id,
                'result': self.result,
                'date_done': self.date_done}

    def __str__(self):
        return '<TaskSet: {0.taskset_id}>'.format(self)


PERIOD_CHOICES = (('days', _('Days')),
                  ('hours', _('Hours')),
                  ('minutes', _('Minutes')),
                  ('seconds', _('Seconds')),
                  ('microseconds', _('Microseconds')))


@python_2_unicode_compatible
class IntervalSchedule(models.Model):
    every = models.IntegerField(_('every'), null=False)
    period = models.CharField(
        _('period'), max_length=24, choices=PERIOD_CHOICES,
    )

    class Meta:
        verbose_name = _('interval')
        verbose_name_plural = _('intervals')
        ordering = ['period', 'every']

    @property
    def schedule(self):
        return schedules.schedule(timedelta(**{self.period: self.every}))

    @classmethod
    def from_schedule(cls, schedule, period='seconds'):
        every = max(schedule.run_every.total_seconds(), 0)
        try:
            return cls.objects.get(every=every, period=period)
        except cls.DoesNotExist:
            return cls(every=every, period=period)
        except MultipleObjectsReturned:
            cls.objects.filter(every=every, period=period).delete()
            return cls(every=every, period=period)

    def __str__(self):
        if self.every == 1:
            return _('every {0.period_singular}').format(self)
        return _('every {0.every:d} {0.period}').format(self)

    @property
    def period_singular(self):
        return self.period[:-1]


@python_2_unicode_compatible
class CrontabSchedule(models.Model):
    minute = models.CharField(_('minute'), max_length=64, default='*')
    hour = models.CharField(_('hour'), max_length=64, default='*')
    day_of_week = models.CharField(
        _('day of week'), max_length=64, default='*',
    )
    day_of_month = models.CharField(
        _('day of month'), max_length=64, default='*',
    )
    month_of_year = models.CharField(
        _('month of year'), max_length=64, default='*',
    )

    class Meta:
        verbose_name = _('crontab')
        verbose_name_plural = _('crontabs')
        ordering = ['month_of_year', 'day_of_month',
                    'day_of_week', 'hour', 'minute']

    def __str__(self):
        return '{0} {1} {2} {3} {4} (m/h/d/dM/MY)'.format(
            cronexp(self.minute),
            cronexp(self.hour),
            cronexp(self.day_of_week),
            cronexp(self.day_of_month),
            cronexp(self.month_of_year),
        )

    @property
    def schedule(self):
        return schedules.crontab(minute=self.minute,
                                 hour=self.hour,
                                 day_of_week=self.day_of_week,
                                 day_of_month=self.day_of_month,
                                 month_of_year=self.month_of_year)

    @classmethod
    def from_schedule(cls, schedule):
        spec = {'minute': schedule._orig_minute,
                'hour': schedule._orig_hour,
                'day_of_week': schedule._orig_day_of_week,
                'day_of_month': schedule._orig_day_of_month,
                'month_of_year': schedule._orig_month_of_year}
        try:
            return cls.objects.get(**spec)
        except cls.DoesNotExist:
            return cls(**spec)
        except MultipleObjectsReturned:
            cls.objects.filter(**spec).delete()
            return cls(**spec)


class PeriodicTasks(models.Model):
    ident = models.SmallIntegerField(default=1, primary_key=True, unique=True)
    last_update = models.DateTimeField(null=False)

    objects = managers.ExtendedManager()

    @classmethod
    def changed(cls, instance, **kwargs):
        if not instance.no_changes:
            cls.objects.update_or_create(ident=1,
                                         defaults={'last_update': now()})

    @classmethod
    def last_change(cls):
        try:
            return cls.objects.get(ident=1).last_update
        except cls.DoesNotExist:
            pass


@python_2_unicode_compatible
class PeriodicTask(models.Model):
    name = models.CharField(
        _('name'), max_length=191, unique=True,
        help_text=_('Useful description'),
    )
    task = models.CharField(_('task name'), max_length=200)
    interval = models.ForeignKey(
        IntervalSchedule,
        null=True, blank=True, verbose_name=_('interval'),
        on_delete=models.CASCADE,
    )
    crontab = models.ForeignKey(
        CrontabSchedule, null=True, blank=True, verbose_name=_('crontab'),
        on_delete=models.CASCADE,
        help_text=_('Use one of interval/crontab'),
    )
    args = models.TextField(
        _('Arguments'), blank=True, default='[]',
        help_text=_('JSON encoded positional arguments'),
    )
    kwargs = models.TextField(
        _('Keyword arguments'), blank=True, default='{}',
        help_text=_('JSON encoded keyword arguments'),
    )
    queue = models.CharField(
        _('queue'), max_length=200, blank=True, null=True, default=None,
        help_text=_('Queue defined in CELERY_QUEUES'),
    )
    exchange = models.CharField(
        _('exchange'), max_length=200, blank=True, null=True, default=None,
    )
    routing_key = models.CharField(
        _('routing key'), max_length=200, blank=True, null=True, default=None,
    )
    expires = models.DateTimeField(
        _('expires'), blank=True, null=True,
    )
    enabled = models.BooleanField(
        _('enabled'), default=True,
    )
    last_run_at = models.DateTimeField(
        auto_now=False, auto_now_add=False,
        editable=False, blank=True, null=True,
    )
    total_run_count = models.PositiveIntegerField(
        default=0, editable=False,
    )
    date_changed = models.DateTimeField(auto_now=True)
    description = models.TextField(_('description'), blank=True)

    objects = managers.PeriodicTaskManager()
    no_changes = False

    class Meta:
        verbose_name = _('periodic task')
        verbose_name_plural = _('periodic tasks')

    def validate_unique(self, *args, **kwargs):
        super(PeriodicTask, self).validate_unique(*args, **kwargs)
        if not self.interval and not self.crontab:
            raise ValidationError(
                {'interval': ['One of interval or crontab must be set.']})
        if self.interval and self.crontab:
            raise ValidationError(
                {'crontab': ['Only one of interval or crontab must be set']})

    def save(self, *args, **kwargs):
        self.exchange = self.exchange or None
        self.routing_key = self.routing_key or None
        self.queue = self.queue or None
        if not self.enabled:
            self.last_run_at = None
        super(PeriodicTask, self).save(*args, **kwargs)

    def __str__(self):
        fmt = '{0.name}: {{no schedule}}'
        if self.interval:
            fmt = '{0.name}: {0.interval}'
        if self.crontab:
            fmt = '{0.name}: {0.crontab}'
        return fmt.format(self)

    @property
    def schedule(self):
        if self.interval:
            return self.interval.schedule
        if self.crontab:
            return self.crontab.schedule


signals.pre_delete.connect(PeriodicTasks.changed, sender=PeriodicTask)
signals.pre_save.connect(PeriodicTasks.changed, sender=PeriodicTask)


class WorkerState(models.Model):
    hostname = models.CharField(_('hostname'), max_length=191, unique=True)
    last_heartbeat = models.DateTimeField(_('last heartbeat'), null=True,
                                          db_index=True)

    objects = managers.ExtendedManager()

    class Meta:
        """Model meta-data."""
        verbose_name = _('worker')
        verbose_name_plural = _('workers')
        get_latest_by = 'last_heartbeat'
        ordering = ['-last_heartbeat']

    def __str__(self):
        return self.hostname

    def __repr__(self):
        return '<WorkerState: {0.hostname}>'.format(self)

    def is_alive(self):
        if self.last_heartbeat:
            # Use UTC timestamp if USE_TZ is true, or else use local timestamp
            timestamp = mktime(gmtime()) if settings.USE_TZ else time()
            return timestamp < heartbeat_expires(self.heartbeat_timestamp)
        return False

    @property
    def heartbeat_timestamp(self):
        return mktime(self.last_heartbeat.timetuple())


@python_2_unicode_compatible
class TaskState(models.Model):
    state = models.CharField(
        _('state'), max_length=64, choices=TASK_STATE_CHOICES, db_index=True,
    )
    task_id = models.CharField(_('UUID'), max_length=36, unique=True)
    name = models.CharField(
        _('name'), max_length=191, null=True, db_index=True,
    )
    tstamp = models.DateTimeField(_('event received at'), db_index=True)
    args = models.TextField(_('Arguments'), null=True)
    kwargs = models.TextField(_('Keyword arguments'), null=True)
    eta = models.DateTimeField(_('ETA'), null=True)
    expires = models.DateTimeField(_('expires'), null=True)
    result = models.TextField(_('result'), null=True)
    traceback = models.TextField(_('traceback'), null=True)
    runtime = models.FloatField(
        _('execution time'), null=True,
        help_text=_('in seconds if task succeeded'),
    )
    retries = models.IntegerField(_('number of retries'), default=0)
    worker = models.ForeignKey(
        WorkerState, null=True, verbose_name=_('worker'),
        on_delete=models.CASCADE,
    )
    hidden = models.BooleanField(editable=False, default=False, db_index=True)

    objects = managers.TaskStateManager()

    class Meta:
        """Model meta-data."""
        verbose_name = _('task')
        verbose_name_plural = _('tasks')
        get_latest_by = 'tstamp'
        ordering = ['-tstamp']

    def __str__(self):
        name = self.name or 'UNKNOWN'
        s = '{0.state:<10} {0.task_id:<36} {1}'.format(self, name)
        if self.eta:
            s += ' eta:{0.eta}'.format(self)
        return s

    def __repr__(self):
        return '<TaskState: {0.state} {1}[{0.task_id}] ts:{0.tstamp}>'.format(
            self, self.name or 'UNKNOWN',
        )


djcelery/__init__.py
"""Old django celery integration project."""
# :copyright: (c) 2009 - 2015 by Ask Solem.
# :license:   BSD, see LICENSE for more details.
from __future__ import absolute_import, unicode_literals

import os
import sys

VERSION = (3, 2, 1, '+singleplatform.1')
__version__ = '.'.join(map(str, VERSION[0:3])) + ''.join(VERSION[3:])
__author__ = 'Ask Solem'
__contact__ = 'ask@celeryproject.org'
__homepage__ = 'http://celeryproject.org'
__docformat__ = 'restructuredtext'
__license__ = 'BSD (3 clause)'

# -eof meta-


if sys.version_info[0] == 3:

    def setup_loader():
        os.environ.setdefault(
            'CELERY_LOADER', 'djcelery.loaders.DjangoLoader',
        )

else:

    def setup_loader():  # noqa
        os.environ.setdefault(
            b'CELERY_LOADER', b'djcelery.loaders.DjangoLoader',
        )

from celery import current_app as celery  # noqa


djcelery/migrations/0001_initial.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals

from django.db import models, migrations
import djcelery.picklefield


class Migration(migrations.Migration):

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='CrontabSchedule',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('minute', models.CharField(default='*', max_length=64, verbose_name='minute')),
                ('hour', models.CharField(default='*', max_length=64, verbose_name='hour')),
                ('day_of_week', models.CharField(default='*', max_length=64, verbose_name='day of week')),
                ('day_of_month', models.CharField(default='*', max_length=64, verbose_name='day of month')),
                ('month_of_year', models.CharField(default='*', max_length=64, verbose_name='month of year')),
            ],
            options={
                'ordering': ['month_of_year', 'day_of_month', 'day_of_week', 'hour', 'minute'],
                'verbose_name': 'crontab',
                'verbose_name_plural': 'crontabs',
            },
            bases=(models.Model,),
        ),
        migrations.CreateModel(
            name='IntervalSchedule',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('every', models.IntegerField(verbose_name='every')),
                ('period', models.CharField(max_length=24, verbose_name='period', choices=[('days', 'Days'), ('hours', 'Hours'), ('minutes', 'Minutes'), ('seconds', 'Seconds'), ('microseconds', 'Microseconds')])),
            ],
            options={
                'ordering': ['period', 'every'],
                'verbose_name': 'interval',
                'verbose_name_plural': 'intervals',
            },
            bases=(models.Model,),
        ),
        migrations.CreateModel(
            name='PeriodicTask',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('name', models.CharField(help_text='Useful description', unique=True, max_length=191, verbose_name='name')),
                ('task', models.CharField(max_length=200, verbose_name='task name')),
                ('args', models.TextField(default='[]', help_text='JSON encoded positional arguments', verbose_name='Arguments', blank=True)),
                ('kwargs', models.TextField(default='{}', help_text='JSON encoded keyword arguments', verbose_name='Keyword arguments', blank=True)),
                ('queue', models.CharField(default=None, max_length=200, blank=True, help_text='Queue defined in CELERY_QUEUES', null=True, verbose_name='queue')),
                ('exchange', models.CharField(default=None, max_length=200, null=True, verbose_name='exchange', blank=True)),
                ('routing_key', models.CharField(default=None, max_length=200, null=True, verbose_name='routing key', blank=True)),
                ('expires', models.DateTimeField(null=True, verbose_name='expires', blank=True)),
                ('enabled', models.BooleanField(default=True, verbose_name='enabled')),
                ('last_run_at', models.DateTimeField(null=True, editable=False, blank=True)),
                ('total_run_count', models.PositiveIntegerField(default=0, editable=False)),
                ('date_changed', models.DateTimeField(auto_now=True)),
                ('description', models.TextField(verbose_name='description', blank=True)),
                ('crontab', models.ForeignKey(blank=True, to='djcelery.CrontabSchedule', help_text='Use one of interval/crontab', null=True, verbose_name='crontab', on_delete=models.CASCADE)),
                ('interval', models.ForeignKey(verbose_name='interval', blank=True, to='djcelery.IntervalSchedule', null=True, on_delete=models.CASCADE)),
            ],
            options={
                'verbose_name': 'periodic task',
                'verbose_name_plural': 'periodic tasks',
            },
            bases=(models.Model,),
        ),
        migrations.CreateModel(
            name='PeriodicTasks',
            fields=[
                ('ident', models.SmallIntegerField(default=1, unique=True, serialize=False, primary_key=True)),
                ('last_update', models.DateTimeField()),
            ],
            options={
            },
            bases=(models.Model,),
        ),
        migrations.CreateModel(
            name='TaskMeta',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('task_id', models.CharField(unique=True, max_length=191, verbose_name='task id')),
                ('status', models.CharField(default='PENDING', max_length=50, verbose_name='state', choices=[('FAILURE', 'FAILURE'), ('PENDING', 'PENDING'), ('RECEIVED', 'RECEIVED'), ('RETRY', 'RETRY'), ('REVOKED', 'REVOKED'), ('STARTED', 'STARTED'), ('SUCCESS', 'SUCCESS')])),
                ('result', djcelery.picklefield.PickledObjectField(default=None, null=True, editable=False)),
                ('date_done', models.DateTimeField(auto_now=True, verbose_name='done at')),
                ('traceback', models.TextField(null=True, verbose_name='traceback', blank=True)),
                ('hidden', models.BooleanField(default=False, db_index=True, editable=False)),
                ('meta', djcelery.picklefield.PickledObjectField(default=None, null=True, editable=False)),
            ],
            options={
                'db_table': 'celery_taskmeta',
                'verbose_name': 'task state',
                'verbose_name_plural': 'task states',
            },
            bases=(models.Model,),
        ),
        migrations.CreateModel(
            name='TaskSetMeta',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('taskset_id', models.CharField(unique=True, max_length=191, verbose_name='group id')),
                ('result', djcelery.picklefield.PickledObjectField(editable=False)),
                ('date_done', models.DateTimeField(auto_now=True, verbose_name='created at')),
                ('hidden', models.BooleanField(default=False, db_index=True, editable=False)),
            ],
            options={
                'db_table': 'celery_tasksetmeta',
                'verbose_name': 'saved group result',
                'verbose_name_plural': 'saved group results',
            },
            bases=(models.Model,),
        ),
        migrations.CreateModel(
            name='TaskState',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('state', models.CharField(db_index=True, max_length=64, verbose_name='state', choices=[('FAILURE', 'FAILURE'), ('PENDING', 'PENDING'), ('RECEIVED', 'RECEIVED'), ('RETRY', 'RETRY'), ('REVOKED', 'REVOKED'), ('STARTED', 'STARTED'), ('SUCCESS', 'SUCCESS')])),
                ('task_id', models.CharField(unique=True, max_length=36, verbose_name='UUID')),
                ('name', models.CharField(max_length=191, null=True, verbose_name='name', db_index=True)),
                ('tstamp', models.DateTimeField(verbose_name='event received at', db_index=True)),
                ('args', models.TextField(null=True, verbose_name='Arguments')),
                ('kwargs', models.TextField(null=True, verbose_name='Keyword arguments')),
                ('eta', models.DateTimeField(null=True, verbose_name='ETA')),
                ('expires', models.DateTimeField(null=True, verbose_name='expires')),
                ('result', models.TextField(null=True, verbose_name='result')),
                ('traceback', models.TextField(null=True, verbose_name='traceback')),
                ('runtime', models.FloatField(help_text='in seconds if task succeeded', null=True, verbose_name='execution time')),
                ('retries', models.IntegerField(default=0, verbose_name='number of retries')),
                ('hidden', models.BooleanField(default=False, db_index=True, editable=False)),
            ],
            options={
                'ordering': ['-tstamp'],
                'get_latest_by': 'tstamp',
                'verbose_name': 'task',
                'verbose_name_plural': 'tasks',
            },
            bases=(models.Model,),
        ),
        migrations.CreateModel(
            name='WorkerState',
            fields=[
                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
                ('hostname', models.CharField(unique=True, max_length=191, verbose_name='hostname')),
                ('last_heartbeat', models.DateTimeField(null=True, verbose_name='last heartbeat', db_index=True)),
            ],
            options={
                'ordering': ['-last_heartbeat'],
                'get_latest_by': 'last_heartbeat',
                'verbose_name': 'worker',
                'verbose_name_plural': 'workers',
            },
            bases=(models.Model,),
        ),
        migrations.AddField(
            model_name='taskstate',
            name='worker',
            field=models.ForeignKey(verbose_name='worker', to='djcelery.WorkerState', null=True, on_delete=models.CASCADE),
            preserve_default=True,
        ),
    ]
